src/ccs/cccs.rs

```diff
@@ -28,47 +28,64 @@ use std::sync::Arc;
 
 use super::util::compute_sum_Mz;
 use super::util::virtual_poly::VirtualPolynomial;
-use super::{CCSShape, CCSWitness};
+use super::CCS;
 
 /// A type that holds the shape of a Committed CCS (CCCS) instance
-#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
-#[serde(bound = "")]
-pub struct CCCSShape<G: Group> {
-  // Sequence of sparse MLE polynomials in s+s' variables M_MLE1, ..., M_MLEt
-  pub(crate) M_MLE: Vec<MultilinearPolynomial<G::Scalar>>,
-
-  pub(crate) ccs: CCSShape<G>,
+#[derive(Clone, Debug, PartialEq, Eq /*Serialize, Deserialize*/)]
+//#[serde(bound(deserialize = "'de: 'a"))]
+pub struct CCCS<G: Group> {
+  // The `z` vector used as input for this instance.
+  pub(crate) z: Vec<G::Scalar>,
+  // Commitment to the witness of `z`.
+  pub(crate) w_comm: Commitment<G>,
+  // Commitment to the public inputs of `z`.
+  pub(crate) x_comm: Commitment<G>,
 }
 
-/// A type that holds a CCCS instance
-#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
-#[serde(bound = "")]
-pub struct CCCSInstance<G: Group> {
-  // Commitment to a multilinear polynomial in s' - 1 variables
-  pub(crate) C: Commitment<G>,
-
-  // $x in F^l$
-  pub(crate) x: Vec<G::Scalar>,
-}
+impl<G: Group> CCCS<G> {
+  /// Generates a new CCCS given a reference to it's original CCS repr & the multilinear extension of it's matrixes.
+  /// Then, given the input vector `z`.
+  pub(crate) fn new(
+    ccs: &CCS<G>,
+    ccs_matrix_mle: &Vec<MultilinearPolynomial<G::Scalar>>,
+    z: Vec<G::Scalar>,
+    ck: &CommitmentKey<G>,
+  ) -> Self {
+    let x_comm = if ccs.l != 0 {
+      CE::<G>::commit(ck, &z[1..ccs.l + 1])
+    } else {
+      Commitment::<G>::default()
+    };
+    let w_comm = CE::<G>::commit(ck, &z[(1 + ccs.l)..]);
+
+    Self {
+      z: z.to_vec(),
+      w_comm,
+      x_comm,
+    }
+  }
 
-impl<G: Group> CCCSShape<G> {
-  // Computes q(x) = \sum^q c_i * \prod_{j \in S_i} ( \sum_{y \in {0,1}^s'} M_j(x, y) * z(y) )
-  // polynomial over x
-  pub fn compute_q(&self, z: &Vec<G::Scalar>) -> Result<VirtualPolynomial<G::Scalar>, NovaError> {
-    let z_mle = dense_vec_to_mle::<G::Scalar>(self.ccs.s_prime, z);
-    if z_mle.get_num_vars() != self.ccs.s_prime {
+  /// Computes q(x) = \sum^q c_i * \prod_{j \in S_i} ( \sum_{y \in {0,1}^s'} M_j(x, y) * z(y) )
+  /// polynomial over x
+  pub(crate) fn compute_q(
+    &self,
+    ccs: &CCS<G>,
```

Comment:
What is the advantage of passing ccs to compute_q vs keeping a reference to it? (Maybe obvious reading code, just leaving notes as I go)

Reply:
Not sure I follow. 
But if it's regarding why `CCCS` doesn't contain a `CCS` inside, it's beucause (as was discussed a few weeks ago) the API and code will get worse if we force structs to hold references inside. Leading to lifetime annotations in a lot of places.